apollo-router/src/plugins/connectors/handle_responses.rs

```diff
@@ -145,28 +140,46 @@ pub(crate) async fn process_response<T: HttpBody>(
             // If this errors, it will write to the debug context because it
             // has access to the raw bytes, so we can't write to it again
             // in any RawResponse::Error branches.
-            let raw = match deserialized_body {
-                Ok(data) => RawResponse::Data {
-                    parts,
-                    data,
-                    key: response_key,
-                    debug_request,
-                },
-                Err(error) => RawResponse::Error {
-                    error,
+            let mapped = match &deserialized_body {
+                Err(error) => MappedResponse::Error {
+                    error: error.clone(),
                     key: response_key,
+                    problems: Vec::new(),
                 },
+                Ok(data) => handle_raw_response(
+                    data,
+                    &parts,
+                    response_key,
+                    &connector,
+                    context,
+                    supergraph_request.headers(),
+                ),
             };
 
-            let (mapped, is_success) = handle_raw_response(
-                raw,
-                &connector,
-                context,
-                debug_context,
-                supergraph_request.headers(),
-            );
-
-            if is_success {
+            if let Some(debug) = debug_context {
+                let mut debug_problems: Vec<Problem> = mapped.problems().to_vec();
+                debug_problems.extend(debug_request.1);
+
+                let selection_data = if let MappedResponse::Data { key, data, .. } = &mapped {
```

Comment:
maybe this could become a function in `MappedResponse`, something like `map_data`:

```
fn map_selection_data(self) -> Option<SelectionData> {
  if let MappedResponse::Data { key, data, .. } = &mapped {
    Some(SelectionData {
        source: connector.selection.to_string(),
        transformed: key.selection().to_string(),
        result: Some(data.clone()),
    })
  } else {
    None
  }
}
```